### PR TITLE
Add reporting and quote hiding

### LINKED ISSUE #47: Can't delete a quote
would be cool if you could delete a quote that you submitted (or maybe additionally a quote that you were quoted saying)

### LINKED ISSUE #58: No database revisioning
Currently, if a change is made to the database schema, a user with administrative access must go in and manually make the changes. 

It’d be cool if we had a tool like Alembic to manage db revisions. This has a couple of really handy features:
* Tracked revisions
* Easily managed upgrades and downgrades
* Easy provisioning of testing instances.

Any revision files would need to be mirrored to https://github.com/ComputerScienceHouse/QuotefaultAPI to make sure they stay in sync.

See also ComputerScienceHouse/QuotefaultAPI#56

### PR DESCRIPTION
**Pull request by mxmeinhold**
Defines database migrations, and adds report functionality and quote hiding.

Requirements:
- [x] Define database migrations and updated schema
- [x] Add quote hiding
- [x] Add report functionality
- [x] Add admin interface
- [x] Add quote hiding button on the storage page for: speaker, submitter, eboard, rtp

Nice to have:
- [x] Notify of new reports by email or by slack

Resolves #58, resolves #47

### CODE REVIEW

**Comment by jabbate19** [comment]
Upon a report, who would it go to? Is there a single email for this to go to or would it have to be sent to all individual RTP/Eboard emails?

**Comment by mxmeinhold** [comment]
I believe eboard@ would work. I can check once power is back.

File: quotefault/mail.py [hunk 1]
```diff
@@ -1,15 +1,39 @@
...
+def send_email(toaddr, subject, body):
+    """
+    Helper function for quote notification email
+    """
+    fromaddr = app.config['MAIL_USERNAME']
     msg = MIMEMultipart()
     msg['From'] = fromaddr
     msg['To'] = toaddr
     msg['Subject'] = subject
-    body = body
```
Comment (mxmeinhold) [inline quotefault/mail.py:12]: Why is this deleted?

File: quotefault/ldap.py [hunk 2]
```diff
@@ -1,24 +1,52 @@
...
         try:
             member = ldap_get_member(username)
             return member.cn + " (" + member.uid + ")"
         except:
             return username
     return dict(get_display_name=get_display_name)
+
+def is_member_of_group(uid: str, group: str) -> bool:
+    """
+    Determine if user is member of OIDC group
```
Comment (mxmeinhold) [inline quotefault/ldap.py:45]: Not oidc, this checks ldap

File: quotefault/mail.py (hunk 1 above)
Comment (jabbate19) [reply quotefault/mail.py:12]: It's body = body, it did nothing

File: quotefault/ldap.py (hunk 2 above)
Comment (jabbate19) [reply quotefault/ldap.py:45]: I can fix up that tmrw

File: config.sample.py [hunk 3]
```diff
@@ -20,5 +20,8 @@
 # CSH_LDAP credentials
 LDAP_BIND_DN = ''
 LDAP_BIND_PW = ''
+MAIL_USERNAME = ''
 MAIL_PASSWORD = ''
 MAIL_SERVER = ''
+MAIL_PORT = 465
+MAIL_USE_SSL = True
```
Comment (Mstrodl) [inline config.sample.py:27]: Can we get consistent EOFs?

File: migrations/versions/3b8a4c7fbcc2_report_reviewed_column_added.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+depends_on = None
+
+
+def upgrade():
+    op.add_column('report', sa.Column('reviewed', sa.Boolean(), nullable=False))
+
+
+def downgrade():
+    op.drop_column('report', 'reviewed')
+
```
Comment (Mstrodl) [inline migrations/versions/3b8a4c7fbcc2_report_reviewed_column_added.py:25]: Double newline? Huh?

File: quotefault/templates/bootstrap/hidden.html [hunk 5]
```diff
@@ -0,0 +1,100 @@
+{% extends "bootstrap/base.html" %}
+
+{% block styles %}
+    <link rel="stylesheet" href="/static/css/votes/upvote.css">
+{% endblock %}
+
+{% block body %}
+    <script type="text/javascript">
+        function unhide(id){
```
Comment (Mstrodl) [inline quotefault/templates/bootstrap/hidden.html:9]: :cry: consistent indentation?

File: quotefault/templates/bootstrap/storage.html [hunk 6]
```diff
@@ -55,4 +151,4 @@
     <script src="{{ url_for('static', filename='js/load_more.js') }}"></script>
     <script src="{{ url_for('static', filename='css/votes/upvote.js') }}"></script>
     <script src="{{ url_for('static', filename='js/vote.js') }}"></script>
-{% endblock %}
+{% endblock %}
```
Comment (Mstrodl) [inline quotefault/templates/bootstrap/storage.html:154]: :pleading_face: line ending

File: requirements.txt [hunk 7]
```diff
@@ -2,10 +2,11 @@ alembic
...
 Flask
+Flask-Mail==0.9.1
 Flask-Migrate
 Flask-pyoidc==1.3.0
 Flask-Script
-Flask-SQLAlchemy
+Flask-SQLAlchemy>=2.5.0
 gunicorn
 itsdangerous
 Jinja2
```
Comment (Mstrodl) [inline requirements.txt:12]: can we pin all of these?? :pleading_face:

**Review by Mstrodl — APPROVED** [review]
LGTM, look at my comments tho

**Comment by Mstrodl** [comment]
@jabbate19 can we get some screenshots??

**Comment by jabbate19** [comment]
I accidentally deleted my config.py with all the creds in it, so I can't really load anything to take screenshots

File: quotefault/ldap.py [hunk 8]
```diff
@@ -1,24 +1,52 @@
+"""
+File name: mail.py
```
Comment (WillNilges) [inline quotefault/ldap.py:2]: File name doesn't match?

File: quotefault/mail.py (hunk 1 above)
Comment (WillNilges) [reply quotefault/mail.py:12]: Is it supposed to be, like, `msg['Body'] = body` or something?

**Review by WillNilges — APPROVED** [review]
LGTM. I have a few documentation-oriented things, but I didn't see anything _egregious._ :upside_down_face:

File: quotefault/mail.py (hunk 1 above)
Comment (jabbate19) [reply quotefault/mail.py:12]: body is passed as its own parameter

File: config.sample.py (hunk 3 above)
Comment (jabbate19) [reply config.sample.py:27]: EOF?

File: quotefault/__init__.py [hunk 9]
```diff
@@ -215,28 +239,138 @@ def get():
...
-    elif metadata['submitter'] is not None:
-        quotes = db.session.query(Quote, func.sum(Vote.direction).label('votes')).outerjoin(Vote).group_by(Quote).order_by(Quote.quote_time.desc()).filter(Quote.submitter == metadata['submitter']).all()
-    elif metadata['speaker'] is not None:
-        quotes = db.session.query(Quote, func.sum(Vote.direction).label('votes')).outerjoin(Vote).group_by(Quote).order_by(Quote.quote_time.desc()).filter(Quote.speaker == metadata['speaker']).all()
-    else:
-        quotes = db.session.query(Quote, func.sum(Vote.direction).label('votes')).outerjoin(Vote).group_by(Quote).order_by(Quote.quote_time.desc()).all()
+
+    # Get all the quotes
+    quotes = get_quote_query(speaker = request.args.get('speaker'),
+        submitter = request.args.get('submitter')).all()
```
[the remark this answers is not in the document]
Comment (mxmeinhold) [reply quotefault/__init__.py:250]: This shouldn't be that slow. We should add datadog db monitoring if we're curious.

Regardless, paginating that is out of scope of this pr.

File: migrations/env.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+# Interpret the config file for Python logging.
+# This line sets up loggers basically.
+fileConfig(config.config_file_name)
+logger = logging.getLogger('alembic.env')
+
+# add your model's MetaData object here
+# for 'autogenerate' support
+# from myapp import mymodel
+# target_metadata = mymodel.Base.metadata
```
[the remark this answers is not in the document]
Comment (jabbate19) [reply migrations/env.py:22]: I env.py gets auto-generated in migrations by alembic. That could be needed at a future time, and if you needed it you would uncomment the lines

[the remark this answers is not in the document]
Comment (mxmeinhold) [reply migrations/env.py:22]: Pretty sure this is part of the autogenerated bits.

File: quotefault/ldap.py (hunk 8 above)
Comment (jabbate19) [reply quotefault/ldap.py:2]: Whoops. I'll fix that rq

File: config.sample.py (hunk 3 above)
Comment (WillNilges) [reply config.sample.py:27]: End Of File. Either have a newline at the end of every file, or don't.

Right @Mstrodl ?

File: alembic.ini [hunk 11]
```diff
@@ -0,0 +1,74 @@
+# A generic, single database configuration.
```
Comment (mxmeinhold) [inline alembic.ini:1]: Do we need this file? I think this wasn't generated by flask migrate cc @jabbate19

File: quotefault/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (WillNilges) [reply quotefault/__init__.py:250]: True, valid, and fair. I just noticed it was on the `diff.`

File: migrations/versions/5c78b9d1d06e_add_report_reviewed_column.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    pass
+
+
+def downgrade():
+    pass
```
Comment (mxmeinhold) [inline migrations/versions/5c78b9d1d06e_add_report_reviewed_column.py:24]: This does nothing.

File: quotefault/__init__.py [hunk 13]
```diff
@@ -1,23 +1,38 @@
...
 from datetime import datetime
 
 import requests
 from csh_ldap import CSHLDAP
 from flask import Flask, render_template, request, flash, session, make_response
+from flask_migrate import Migrate
 from flask_pyoidc.flask_pyoidc import OIDCAuthentication
 from flask_sqlalchemy import SQLAlchemy
 from sqlalchemy import func
+from sqlalchemy.orm.query import Query
```
Comment (mxmeinhold) [inline quotefault/__init__.py:17]: Where is this referenced?

File: quotefault/__init__.py [hunk 14]
```diff
@@ -1,23 +1,38 @@
...
+from werkzeug.utils import redirect
+
+from config import LDAP_BIND_DN
```
Comment (mxmeinhold) [inline quotefault/__init__.py]: Why is this imported instead of using app.config syntax?

File: migrations/versions/3b8a4c7fbcc2_report_reviewed_column_added.py (hunk 4 above)
Comment (jabbate19) [reply migrations/versions/3b8a4c7fbcc2_report_reviewed_column_added.py:25]: Auto-generated formatting by alembic

File: quotefault/__init__.py [hunk 15]
```diff
@@ -28,69 +43,39 @@
...
-        "plug": plug
+        "plug": plug,
+        "is_admin" : is_member_of_group(uid,'eboard') or is_member_of_group(uid,'rtp')
     }
     return metadata
 
-
-# run the main page by creating the table(s) in the CSH serverspace and rendering the mainpage template
+# run the main page by creating the table(s)
+# in the CSH serverspace and rendering the mainpage template
```
Comment (mxmeinhold) [inline quotefault/__init__.py:70]: this function no longer crates tables

File: quotefault/templates/bootstrap/hidden.html (hunk 5 above)
Comment (jabbate19) [reply quotefault/templates/bootstrap/hidden.html:9]: I could be blind but I don't see what's inconsistent here

File: quotefault/__init__.py [hunk 16]
```diff
@@ -215,28 +239,138 @@ def get():
...
 
-
     return render_template(
         'bootstrap/additional_quotes.html',
         quotes=quotes[20:],
         metadata=metadata,
         user_votes=user_votes
     )
+
+@app.route('/report/<report_id>', methods=['POST'])
```
Comment (mxmeinhold) [inline quotefault/__init__.py:260]: I'd prefer this be `quote_id`

File: quotefault/__init__.py [hunk 17]
```diff
@@ -215,28 +239,138 @@ def get():
...
+        flash("You already submitted a report for this Quote!")
+        return redirect('/storage')
+    new_report = Report(report_id, metadata['uid'], None)
+    db.session.add(new_report)
+    db.session.commit()
+    if app.config['MAIL_SERVER'] != '':
+        send_report_email( metadata['uid'], Quote.query.get(report_id) )
+
+    flash("Report Successful!")
+    return redirect('/storage')
```
Comment (mxmeinhold) [inline quotefault/__init__.py:279]: If the user loaded the page with query params, will this redirect lose them?

File: quotefault/__init__.py [hunk 18]
```diff
@@ -215,28 +239,138 @@ def get():
...
+    if metadata['is_admin']:
+        # Get the most recent 20 quotes
+        reports = Report.query.filter(Report.reviewed == False).all()
+
+        return render_template(
+            'bootstrap/admin.html',
+            reports=reports,
+            metadata=metadata
+        )
+    return redirect('/')
```
Comment (mxmeinhold) [inline quotefault/__init__.py:298]: Shouldn't this be a 403 or something?

File: quotefault/__init__.py [hunk 19]
```diff
@@ -215,28 +237,138 @@ def get():
...
+@app.route('/review/<report_id>/<result>', methods=['POST'])
+@auth.oidc_auth
+def review_submit(report_id, result):
+    """
+    Called when Admin decides on a quote being hidden or kept
+    """
+    metadata = get_metadata()
+    result = int(result)
+    if metadata['is_admin']:
+        # 1 = Keep, 0 = Hide
```
Comment (mxmeinhold) [inline quotefault/__init__.py:309]: Why is this an int? why not just use strings?

File: quotefault/__init__.py [hunk 20]
```diff
@@ -215,28 +237,138 @@ def get():
...
+    quote = Quote.query.get(report_id)
+    if ( metadata['uid'] == quote.speaker
+        or metadata['uid'] == quote.submitter
+        or metadata['is_admin'] ):
+        quote.hidden = True
+        db.session.commit()
+        flash("Quote Hidden!")
+    return redirect('/storage')
+
+@app.route('/unhide/<report_id>', methods=['POST'])
```
Comment (mxmeinhold) [inline quotefault/__init__.py:340]: same questions about redirects, plus you should return an error if the user isn't allowed.

**Comment by jabbate19** [comment]
All corrections except for the pinning versions should be corrected as of [5685680](https://github.com/ComputerScienceHouse/Quotefault/pull/60/commits/5685680e267bfbb476ee9e8e1d077f0a1713a09b)

File: quotefault/__init__.py [hunk 21]
```diff
@@ -215,28 +237,138 @@ def get():
...
+    metadata = get_metadata()
+    if metadata['is_admin']:
+        quote = Quote.query.get(report_id)
+        quote.hidden = False
+        db.session.commit()
+        flash("Quote Unhidden!")
+        return redirect('/hidden')
+    return redirect('/')
+
+@app.route('/hidden', methods=['GET'])
```
Comment (mxmeinhold) [inline quotefault/__init__.py:355]: ditto: redirects and 403s

File: quotefault/__init__.py [hunk 22]
```diff
@@ -215,28 +237,138 @@ def get():
...
+    """
+    Presents hidden quotes to qualified users
+    Users who submitted a hidden quote or were quoted in a hidden quote will see those only
+    Admins see all hidden quotes
+    """
+    metadata = get_metadata()
+    if metadata['is_admin']:
+        quotes = Quote.query.filter( Quote.hidden ).all()
+    else:
+        quotes = Quote.query.filter(
```
Comment (mxmeinhold) [inline quotefault/__init__.py:367]: You should use the get quote query function.

**Review by mxmeinhold — COMMENTED** [review]
Please consider this to be requesting changes. I can't leave review statuses since this is my PR. 

cc @Mstrodl @WillNilges please don't merge this

File: quotefault/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mxmeinhold) [reply quotefault/__init__.py:248]: Yeah, it was a refactor. Operation didn't change though.

File: quotefault/__init__.py (hunk 14 above)
Comment (jabbate19) [reply quotefault/__init__.py]: This was fixed in one of my later commits

File: config.sample.py (hunk 3 above)
Comment (mxmeinhold) [reply config.sample.py]: Please have a newline at the end of every file. This is part of the unix definition of a file and expected by a number of tools.

**Review by WillNilges — CHANGES_REQUESTED** [review]
@jabbate19 Please look at and correct @mxmeinhold's comments.

File: quotefault/__init__.py (hunk 17 above)
Comment (jabbate19) [reply quotefault/__init__.py:279]: Yes

File: migrations/README [hunk 23]
```diff
@@ -0,0 +1,10 @@
+Generic single-database configuration.
+
+Q: What is Alembic/Migrations?
+A: Migrations give a versioned history of changes to the SQL Schema for this project. A "revision" defines changes to a table (add/drop tables or columns). This allows you to go back in time if needed, and quickly re-create the whole schema for a local development
+
+Q: How do I update my local database to use the migrations?
+A: (1) Add your URI to config.py or config.env.py, depending on how you are doing it. (2) Run "flask db upgrade"
```
Comment (mxmeinhold) [inline migrations/README:7]: People shouldn't put secrets in `config.env.py`, so let's not suggest they do?

File: migrations/README [hunk 24]
```diff
@@ -0,0 +1,10 @@
...
+
+Q: I made changes to the DB and need a new revision. How do I do that?
```
Comment (mxmeinhold) [inline migrations/README:9]:
```suggestion
Q: I made changes to the models file and need a new revision. How do I do that?
```

File: quotefault/__init__.py (hunk 18 above)
Comment (jabbate19) [reply quotefault/__init__.py:298]: I would still want it to redirect instead of just crashing because it looks better, would this be a time to use `redirect('/',403)`?

File: quotefault/__init__.py (hunk 19 above)
Comment (jabbate19) [reply quotefault/__init__.py:309]: I felt the binary method would be simpler because its only 1 character

File: quotefault/__init__.py (hunk 17 above)
Comment (mxmeinhold) [reply quotefault/__init__.py:279]: 😿 is there a way we could make this not annoying for users? Page reloads aren't the nicest, specially when you might have to scroll a lot to get to where you were.

File: quotefault/__init__.py (hunk 18 above)
Comment (mxmeinhold) [reply quotefault/__init__.py:298]: I mean, it doesn't have to crash, we can give a nice error page.

File: quotefault/__init__.py (hunk 19 above)
Comment (mxmeinhold) [reply quotefault/__init__.py:309]: Yeah but now you're doing int parsing and it's less clear what to send to hide vs not hide.

File: quotefault/__init__.py (hunk 18 above)
Comment (jabbate19) [reply quotefault/__init__.py:298]: I think for now, I can make it redirect to /, use a 403, and maybe flash a "Hey, you're not supposed to go there!" message

File: quotefault/__init__.py (hunk 22 above)
Comment (jabbate19) [reply quotefault/__init__.py:367]: get quote query wouldn't cover the function of the hidden. It can add hidden to the query or remove them, can't show only hidden. Also can look for a specific speaker and submitter, but I would need if a specific person is either one. If anything I'd re-write a similar function below it

Comment (mxmeinhold) [reply quotefault/__init__.py:367]: get quote query returns a query, you can add filtering on top of it.

File: quotefault/__init__.py (hunk 18 above)
Comment (mxmeinhold) [reply quotefault/__init__.py:298]: Okay but that should be implemented as a flask error handler I think

File: quotefault/__init__.py (hunk 22 above)
Comment (mxmeinhold) [reply quotefault/__init__.py:367]: It's meant to handle the common denominator, so we can have consistent access. If it really doesn't fit your use case, makes sense, but if it will work with an additional optional param, or as is, I'd prefer to use the helper.

File: quotefault/__init__.py [hunk 25]
```diff
@@ -1,23 +1,34 @@
...
 else:
     app.config.from_pyfile(os.path.join(os.getcwd(), "config.env.py"))
+
 #var representing the quote database the app is connected to
 db = SQLAlchemy(app)
+migrate = Migrate(app, db)
+app.logger.info('SQLAlchemy pointed at ' + repr(db.engine.url))
+
+# pylint: disable=wrong-import-position
+from .models import Quote, Vote, Report
```
Comment (mxmeinhold) [inline quotefault/__init__.py:30]: Move this down to the `.ldap` and `.models` imports maybe?

File: quotefault/__init__.py [hunk 26]
```diff
@@ -28,69 +39,38 @@
...
+            ['git', 'rev-parse', '--short', 'HEAD']
+        ).decode('utf-8').rstrip()
     plug = request.cookies.get('plug')
     metadata = {
         "uuid": uuid,
         "uid": uid,
         "version": version,
-        "plug": plug
+        "plug": plug,
+        "is_admin" : is_member_of_group(uid,'eboard') or is_member_of_group(uid,'rtp')
```
Comment (mxmeinhold) [inline quotefault/__init__.py:62]:
```suggestion
        "is_admin" : is_member_of_group(uid, 'eboard') or is_member_of_group(uid, 'rtp')

```

File: quotefault/__init__.py (hunk 15 above)
Comment (mxmeinhold) [reply quotefault/__init__.py]: The new version doesn't make sense @jabbate19. What does `in the CSH serverspace and rendering the mainpage template` mean?

File: quotefault/__init__.py [hunk 27]
```diff
@@ -304,53 +299,60 @@ def review_submit(report_id, result):
...
             report.reviewed = True
             db.session.commit()
             flash("Report Completed: Quote Hidden")
+        else:
+            abort(400)
         return redirect('/review')
-    return redirect('/')
+    elif report:
+        abort(403)
+    abort(400)
```
Comment (mxmeinhold) [inline quotefault/__init__.py]: This is kinda confusing. Instead of the nested ifs, could you do:
```
if not admin:
	403

if not report:
	404

if hide:
elif keep:
else:
```

File: quotefault/__init__.py [hunk 28]
```diff
@@ -304,53 +299,60 @@ def review_submit(report_id, result):
...
         or metadata['uid'] == quote.submitter
         or metadata['is_admin'] ):
         quote.hidden = True
         db.session.commit()
         flash("Quote Hidden!")
-    return redirect('/storage')
+        return redirect('/storage')
+    elif quote:
+        abort(403)
+    abort(400)
```
Comment (mxmeinhold) [inline quotefault/__init__.py]: Same as above. Reduce nesting, return 403, then 404, then results.
Also, this should be a 404 instead of a 400.

File: quotefault/__init__.py [hunk 29]
```diff
@@ -304,53 +299,60 @@ def review_submit(report_id, result):
...
+    quote = Quote.query.get(quote_id)
+    if quote and metadata['is_admin']:
         quote.hidden = False
         db.session.commit()
         flash("Quote Unhidden!")
         return redirect('/hidden')
-    return redirect('/')
+    elif quote:
+        abort(403)
+    abort(400)
```
Comment (mxmeinhold) [inline quotefault/__init__.py]: Ditto. Reduce nesting, and that should be a 404.

File: quotefault/__init__.py [hunk 30]
```diff
@@ -362,13 +364,21 @@ def hidden():
...
         metadata=metadata
     )
+
+@app.errorhandler(403)
+def forbidden(e):
+    return render_template('bootstrap/403.html', metadata=get_metadata()), 403
+
+@app.errorhandler(400)
+def forbidden(e):
+    return render_template('bootstrap/400.html', metadata=get_metadata()), 400
```
Comment (mxmeinhold) [inline quotefault/__init__.py:384]: Missing newline at the end of the file.

File: quotefault/templates/bootstrap/admin.html [hunk 31]
```diff
@@ -32,14 +32,14 @@
...
                 <div class="card-footer">
                     Submitted by <a href="https://profiles.csh.rit.edu/user/{{ report.quote.submitter }}">{{ get_display_name(report.quote.submitter) }}</a> on {{ report.quote.quote_time.strftime('%Y-%m-%d %H:%M:%S') }}
-                    <form method="POST" action="/review/{{report.id}}/0" id="hide::{{report.id}}" method="POST">
+                    <form method="POST" action="/review/{{report.id}}/hide" id="hide::{{report.id}}" method="POST">
                         <button type="submit" class="btn btn-danger float-right">Hide</button>
                     </form>
-                    <form method="POST" action="/review/{{report.id}}/1" id="keep::{{report.id}}" method="POST">
+                    <form method="POST" action="/review/{{report.id}}/keep" id="keep::{{report.id}}" method="POST">
                         <button type="submit" class="btn btn-success float-right">Keep</button>
                     </form>
```
Comment (mxmeinhold) [inline quotefault/templates/bootstrap/admin.html]: Hmm. I wonder if buttons and calling out to js would be better than forms. It would also allow for better error handling and we could avoid reloading the page to update results. I'm undecided whether this is something I want to push for in this PR or something we can just push to an issue, thoughts? cc @Mstrodl @WillNilges

Comment (jabbate19) [reply quotefault/templates/bootstrap/admin.html]: All issues fixed except for this, imo it seems to good for now but up to you all

File: migrations/versions/3b8a4c7fbcc2_report_reviewed_column_added.py (hunk 4 above)
Comment (Mstrodl) [reply migrations/versions/3b8a4c7fbcc2_report_reviewed_column_added.py]: Still do it...

File: quotefault/templates/bootstrap/hidden.html (hunk 5 above)
Comment (Mstrodl) [reply quotefault/templates/bootstrap/hidden.html:9]: the `function` declaraction and `var modal = ...` are at the same level... that's bad!

File: quotefault/templates/bootstrap/admin.html (hunk 31 above)
Comment (WillNilges) [reply quotefault/templates/bootstrap/admin.html]: I do like the idea of avoiding a reload. Lord knows we have enough of that in CSH services. How much work would it be?

Comment (mxmeinhold) [reply quotefault/templates/bootstrap/admin.html]: A little JS, maybe a tiny change to the html template too.

Comment (WillNilges) [reply quotefault/templates/bootstrap/admin.html]: Personally, I'm cool with this being a roadmap feature. We have a solid first implementation and people are eager to get quotefault back :upside_down_face:

Comment (Mstrodl) [reply quotefault/templates/bootstrap/admin.html]: This shouldn't be super hard, I think you should go for it. If you want help with this let me know...

File: quotefault/__init__.py [hunk 32]
```diff
@@ -215,28 +230,152 @@ def get():
...
+@app.route('/review', methods=['GET'])
+@auth.oidc_auth
+def review():
+    """
+    Presents page for admins to review reports
+    and either hide or keep quote
+    """
+    metadata = get_metadata()
+    if metadata['is_admin']:
+        # Get the most recent 20 quotes
```
Comment (mxmeinhold) [inline quotefault/__init__.py:282]: quotes? there's also no limit on the query, so the 20 is wrong too

File: quotefault/__init__.py [hunk 33]
```diff
@@ -215,28 +230,152 @@ def get():
...
+        report.reviewed = True
+        db.session.commit()
+        flash("Report Completed: Quote Kept")
+    elif result == "hide":
+        report.quote.hidden = True
+        report.reviewed = True
+        db.session.commit()
+        flash("Report Completed: Quote Hidden")
+    else:
+        abort(404)
```
Comment (mxmeinhold) [inline quotefault/__init__.py:314]:
```suggestion
        abort(400)
```
 404 is not found, 400 is bad request. Prefer 400 for invalid data, like a bad arg, and 404 for missing records, like "there's no report by that id"

File: quotefault/__init__.py [hunk 34]
```diff
@@ -215,28 +230,152 @@ def get():
...
+    abort(403)
+
+@app.route('/review/<report_id>/<result>', methods=['POST'])
+@auth.oidc_auth
+def review_submit(report_id, result):
+    """
+    Called when Admin decides on a quote being hidden or kept
+    """
+    metadata = get_metadata()
+    report = Report.query.get(report_id)
```
Comment (mxmeinhold) [inline quotefault/__init__.py:299]: What happens if someone reviews a report that was already reviewed?

File: quotefault/templates/mail/report.html [hunk 35]
```diff
@@ -0,0 +1,12 @@
+{% extends "extend/email.html" %}
+
+{% block body %}
+    <div class="container main">
+        <h2>Hello EBoard/RTP,</h2>
+        <h3>There is a new report on QuoteFault</h3>
+        <p>Report submitted by: {{reporter}}</p>
+        <p>Quote being reported: {{quote.quote}} -{{quote.speaker}}</p>
+        <p>Quote submitted by: {{quote.submitter}}</p>
+        <p>Please attend to this on https://quotefault.csh.rit.edu/admin</p>
```
Comment (mxmeinhold) [inline quotefault/templates/mail/report.html:10]: Does this resolve to a url in email clients, or do you have to wrap it in an `a` tag?

File: quotefault/__init__.py (hunk 33 above)
Comment (jabbate19) [reply quotefault/__init__.py:314]: Didn't you tell me to make it 404 over 400?

Comment (jabbate19) [reply quotefault/__init__.py:314]: https://github.com/ComputerScienceHouse/Quotefault/pull/60#discussion_r748648468

Comment (jabbate19) [reply quotefault/__init__.py:314]: Oh, nvm. I got it

File: migrations/README [hunk 36]
```diff
@@ -0,0 +1,10 @@
+Generic single-database configuration.
+
+Q: What is Alembic/Migrations?
+A: Migrations give a versioned history of changes to the SQL Schema for this project. A "revision" defines changes to a table (add/drop tables or columns). This allows you to go back in time if needed, and quickly re-create the whole schema for a local development
+
+Q: How do I update my local database to use the migrations?
+A: (1) Add your URI to config.py, depending on how you are doing it. (2) Run "flask db upgrade"
+
+Q: I made changes to the models.py file and need a new revision. How do I do that?
+A: flask db revision -m "Whatever you want to name it"
```
Comment (mxmeinhold) [inline migrations/README:10]:
```suggestion
## What are Alembic and migrations?
Migrations give a versioned history of changes to the SQL Schema for this project. A `revision` or `migration` defines changes to the structure of the database (such as adding/dropping tables or columns), as well as procedures for migrating existing data to the new structure. Migrations allow for safe(ish) rollbacks, and make it easier to make incremental changes to the database.

## Applying migrations
Once you have your db URI in `config.py`, run `flask db upgrade` to bring the db up to the latest revision.

## Creating new revisions
When you change `models.py`, you'll need to make a new migration. run `flask db revision -m "<descriptive title>"` to autogenerate a revision. You should go manually inspect the created revision to make sure it does what you expect. Once you've got your changes all set, be sure to commit the migration file in the same commit as your changes to `models.py`
```

**Comment by jabbate19** [comment]
https://github.com/ComputerScienceHouse/Quotefault/issues/62
https://github.com/ComputerScienceHouse/Quotefault/issues/61

Follow up issues to this PR for anyone to take

File: quotefault/models.py [hunk 37]
```diff
@@ -0,0 +1,90 @@
...
+    reports = db.relationship('Report', back_populates='quote')
+
+    # initialize a row for the Quote table
+    def __init__(self, submitter, quote, speaker):
+        self.quote_time = datetime.now()
+        self.submitter = submitter
+        self.quote = quote
+        self.speaker = speaker
+
+
```
Comment (WillNilges) [inline quotefault/models.py:33]: Newline >:(

**Review by WillNilges — APPROVED** [review]
:ship:

File: quotefault/models.py (hunk 37 above)
Comment (mxmeinhold) [reply quotefault/models.py:33]: I'm fine with double newlines between classes

Comment (mxmeinhold) [reply quotefault/models.py:33]: Actually, I made these double newlines, so I'm going to mark this resolved
